include/ginkgo/core/factorization/factorization.hpp

```diff
@@ -0,0 +1,165 @@
+     * L has an implicit unit diagonal
+     */
+    combined_lu,
+    /*
+     * the factorization L * D * U is stored as L + D + U - 2I, where
+     * L and U have implicit unit diagonals
+     */
+    combined_ldu,
+    /**
+     * the two factors are stored as a composition L * L^H or L * D * L^H
+     * where L and L^T are Csr matrices and D is a Diagonal matrix
+     */
+    symm_composition,
+    /*
+     * the factorization L * L^H is symmetric and stored as a single matrix
+     * containing L + L^H - diag(L)
+     */
+    symm_combined_cholesky,
+    /*
+     * the factorization is symmetric and stored as a single matrix containing
+     * L + D + L^H - 2 * diag(L), where L and L^H have an implicit unit diagonal
+     */
+    symm_combined_ldl,
+};
+
+
+template <typename ValueType, typename IndexType>
+class Factorization : public EnableLinOp<Factorization<ValueType, IndexType>> {
+    friend class EnablePolymorphicObject<Factorization, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+    using diag_type = matrix::Diagonal<ValueType>;
+    using composition_type = Composition<ValueType>;
+
+    /**
+     * Transforms the factorization from a compact representation suitable only
+     * for triangular solves to a composition representation that can also be
+     * used to access individual factors and multiply with the factorization.
+     *
+     * @return  a new Factorization object containing this factorization
+     *          represented as storage_type::composition.
+     */
+    std::unique_ptr<Factorization> unpack() const;
+
+    storage_type get_storage_type() const;
+
+    std::shared_ptr<const matrix_type> get_lower_factor() const;
+
+    std::shared_ptr<const diag_type> get_diagonal() const;
+
+    std::shared_ptr<const matrix_type> get_upper_factor() const;
+
+    std::shared_ptr<const matrix_type> get_combined() const;
+
+    /** Creates a deep copy of the factorization. */
+    Factorization(const Factorization&);
+
+    /** Moves from the given factorization, leaving it empty. */
+    Factorization(Factorization&&);
+
+    Factorization& operator=(const Factorization&);
+
+    Factorization& operator=(Factorization&&);
+
+    static std::unique_ptr<Factorization> create_from_composition(
+        std::unique_ptr<composition_type>);
+
+    static std::unique_ptr<Factorization> create_from_symm_composition(
+        std::unique_ptr<composition_type>);
+
+    static std::unique_ptr<Factorization> create_from_combined_lu(
+        std::unique_ptr<matrix_type>);
+
+    static std::unique_ptr<Factorization> create_from_combined_ldu(
+        std::unique_ptr<matrix_type>);
+
+    static std::unique_ptr<Factorization> create_from_combined_cholesky(
+        std::unique_ptr<matrix_type>);
+
+    static std::unique_ptr<Factorization> create_from_combined_ldl(
+        std::unique_ptr<matrix_type>);
```

Comment:
This approach is meant to prevent people from creating invalid states, since each storage_type has its own requirements w.r.t. how many LinOps of what type are expected

Reply:
can you elaborate what you meen by constexpr instantiations?

Reply:
having something like
```
constexpr storage_type st_empty = storage_type::empty;
```
I guess I was describing that incorrectly.

Reply:
What is the purpose of these aliases? Users would still need to prefix them with the typename, so they would not shorten the code?

Reply:
I don't think you need typename because it's a variable (but I didn't check it). Also we can just keep it as it is.

Reply:
I meant typename the concept not the keyword ;) You'd need to write `Direct<...>::st_empty` instead of `storage_type::empty`, does that make a difference?

Reply:
I would have put the constants just in the solver namespace, perhaps within another namespace, but not in the Direct class. But again, it doesn't make sense to persue that here.